### ISSUE TITLE
FPS / performance analysis

### ISSUE THREAD

**Issue by TTimo** [evancz/first-person-elm#4]
This is really cool from a technology perspective, however the performance is pretty abysmal.

I tried Chrome and FireFox, OSX and Linux .. albeit very briefly .. it's nowhere near the fps and control response as a some simple OpenGL in a tiny native app. Or even Unity web player.

I know there are a lot of reasons for this, and that's a typical problem with this sort of tech stack, everyone passes the hot potato around .. to JS, to WebGL, to the browser etc.

It would be very interesting if someone with the technical know-how and time was able to profile this and point out where things are stalling, buffered, and generally where time is spent.

**Comment by evancz** [comment]
@johnpmayer can say for sure, but the WebGL bindings are very new and I don't think we've done any formal benchmarking. The API is designed to be good about limiting traffic to gpu (which I suspect is the issue if we are responsible for this) and I can easily imagine that the implementation got some caching wrong or something.

In any case, performance has not been a major focus so far because we had other design challenges, but that means there's quite a lot of easy stuff we can do. This will be one of the focuses this summer, so I very very much support taking a data driven approach and profiling as much as possible.

**Comment by johnpmayer** [comment]
Yeah... I need to learn how to properly profile JavaScript. I have
confirmed that all of the caching is working correctly.

**Comment by johnpmayer** [comment]
Looks like a lot of time is spent recalculating the height of the overlaid
plaintext

**Comment by evancz** [comment]
That makes sense some sense. I wonder if lifting it out or getting rid of it makes a big difference.

I think let-floating can help so constant things are not computed unnecessarily. I've been thinking about some tricks to avoid computing dimensions of text. I mentioned one in #637 that'd let you avoid recomputing big chunks of elements, and this would be a fine candidate for that.

**Comment by evancz** [comment]
I made it so that the text never gets re-rendered. Does that make a difference for you guys?
